neuraxle/steps/data.py

```diff
@@ -0,0 +1,140 @@
+    limitations under the License.
+
+..
+    Thanks to Umaneo Technologies Inc. for their contributions to this Machine Learning
+    project, visit https://www.umaneo.com/ for more information on Umaneo Technologies Inc.
+
+"""
+import random
+from typing import Iterable
+
+from neuraxle.base import BaseStep, MetaStepMixin, NonFittableMixin
+from neuraxle.steps.output_handlers import InputAndOutputTransformerMixin
+
+
+class DataShuffler(NonFittableMixin, InputAndOutputTransformerMixin, BaseStep):
+    """
+    Data Shuffling step that shuffles data inputs, and expected_outputs at the same time.
+
+    .. code-block:: python
+
+        p = DataShuffler(
+            EpochRepeater(
+                ForecastingPipeline(),
+                epochs=EPOCHS
+            ),
+            seed=42,
+            increment_seed_after_each_fit=True
+        )
+
+    .. warning::
+        You probably always want to wrap this step by a :class:`TrainOnlyWrapper`
+
+    .. seealso::
+        :class:`TrainOnlyWrapper`,
+        :class:`InputAndOutputTransformerMixin`,
+        :class:`BaseStep`
+    """
+
+    def __init__(self, seed, increment_seed_after_each_fit):
+        InputAndOutputTransformerMixin.__init__(self)
+        BaseStep.__init__(self)
+        self.seed = seed
+        self.increment_seed_after_each_fit = increment_seed_after_each_fit
+
+    def transform(self, data_inputs):
+        """
+        Shuffle data inputs, and expected outputs.
+
+        :param data_inputs: (data inputs, expected outputs) tuple to shuffle
+        :return:
+        """
+        if self.increment_seed_after_each_fit:
+            self.seed += 1
+
+        di, eo = data_inputs
+        data = list(zip(di, eo))
+        random.Random(self.seed).shuffle(data)
+
+        data_inputs_shuffled, expected_outputs_shuffled = list(zip(*data))
+
+        return list(data_inputs_shuffled), list(expected_outputs_shuffled)
+
+
+class EpochRepeater(MetaStepMixin, BaseStep):
+    """
+    Repeat wrapped step fit, or transform for the number of epochs passed in the constructor.
+
+    .. code-block:: python
+
+        p = EpochRepeater(
+            ForecastingPipeline(),
+            epochs=30
+        )
+
+    .. seealso::
+        :class:`MetaStepMixin`,
+        :class:`BaseStep`
+    """
+
+    def __init__(self, wrapped, epochs, fit_only=True):
+        BaseStep.__init__(self)
+        MetaStepMixin.__init__(self, wrapped)
+        self.fit_only = fit_only
+        self.epochs = epochs
+
+    def fit_transform(self, data_inputs, expected_outputs=None) -> ('BaseStep', Iterable):
+        """
+        Fit transform wrapped step self.epochs times.
+
+        :param data_inputs: data inputs to fit on
+        :param expected_outputs: expected_outputs to fit on
+        :return: fitted self
+        """
+        for _ in range(self.epochs):
+            self.wrapped, data_inputs = self.wrapped.fit_transform(data_inputs, expected_outputs)
```

Comment:
Here, I would rather do `fit` for `for _ in range(self.epochs - 1)`, and then a final `fit_transform` on just the last pass. 

Also, another comment: 
Should the last transform then be de-shuffled? like doing `for _ in range(self.epochs)` (not -1) and then only one transform on original data container? I'm hesitant here.

Reply:
I think we should only use the DataShuffler step in train mode. I added TrainOnlyWrapper inside the code example to make sure. Maybe we should also add a TrainOnlyWrapper on top of EpochRepeter.

Reply:
This way the validation split wrapper will neither shuffle data nor repeat for epochs when predicting the validation split data.